src/Tools/Pagination/Paginator.php

```diff
@@ -205,10 +252,20 @@ private function getCountQuery(): Query
         $countQuery = $this->cloneQuery($this->query);
 
         if (! $countQuery->hasHint(CountWalker::HINT_DISTINCT)) {
-            $countQuery->setHint(CountWalker::HINT_DISTINCT, true);
+            $hintDistinctDefaultTrue = true;
+
+            // When not joining onto *ToMany relations, then use a simpler COUNT query in the CountWalker.
```

Comment:
`fetchJoinCollection` is not only about whether we join on a ToMany relation, but whether we include this relation in what get fetched by the ORM. So the condition you use does not match what the comment says.

Reply:
I see what you mean. It appears I misunderstood the purpose of that variable.

I gave a thought about what to do with it, and I initially concluded that that variable is just misnamed. Consider the following:

1. The `toIterator()`'s main if-statement ([link](https://github.com/doctrine/orm/blob/5b8b5f28f535e1f03b54dcfb0427407ed92f5b72/src/Tools/Pagination/Paginator.php#L115)) executes the extra "SELECT DISTINCT" subquery only if `fetchJoinCollection` is true. Otherwise, the original query is executed as-provided (with an OFFSET and LIMIT attached to it). Note that if `fetchJoinCollection` is supposed to be true only when (a) the original query joins onto a *ToMany relation _AND_ only when (b) the query fetch-loads that relation, then `toInterator()` would _not_ run the extra "SELECT DISTINCT" subquery when the query does the first without the second, which would produce an incorrect result for the "current pagination page". The result would be incorrect because even though the *ToMany relation is not fetch-loaded, the database would still return (and paginate on) "duplicate records" of the entities that are fetch-loaded (which most likely is only the "root" entity).

In other words, both the following DQL queries require the extra "SELECT DISTINCT" subquery to happen in the `toIterator()`, otherwise the result is incorrect:
```
SELECT u FROM User u JOIN u.tags ut;
SELECT u, ut FROM User u JOIN u.tags ut;
```

2. The docblocks for the `fetchJoinCollection` consistently say the following:
```
/** @param bool $fetchJoinCollection Whether the query joins a collection (true by default). */

// (...)

/**
 * Returns whether the query joins a collection.
 *
 * @return bool Whether the query joins a collection.
 */
public function getFetchJoinCollection(): bool {}
```
It appears then that the docblocks are under an impression that that variable is "true" regardless of whether a *ToMany relation is fetch-loaded or not.

---

Do you agree that that instance variable is not named correctly (or the author incorrectly thought that not fetch-loading a *ToMany relation is immune to the "duplicated records being returned from the database" problem)? Would you agree for me to just rename that instance variable to something like: "queryHasToManyJoins"?

Reply:
hmm, maybe it is indeed badly named.

Reply:
Note that there is a case where joined collections that are not selected will not impact the count or the iteration: when using `SELECT DISTINCT` in the query (see `SelectClause::$isDistinct`)

Reply:
Just so that I understand: you're trying to say that running the following DQLs without an "outputWalker" would be fine because it would produce the correct results, therefore the literal meaning of `fetchJoinCollection` is the correct meaning?
```
SELECT DISTINCT u FROM User u JOIN u.userGroups;
SELECT COUNT(DISTINCT u) FROM User u JOIN u.userGroups;
```
I agree, however note that in the current code there is nothing that would ensure that DISTINCT is added/present in the `toInterator()`'s `else` branch of the main if-else-statement. I.e. what you say is correct, but the current code would still either not work, or work by an extreme fluke (i.e. because someone did add a DISTINCT keyword to their `$query` prior to using the Paginator). Based on this, I would still say that `fetchJoinCollection` should just be renamed to reflect the fact that it's actual present meaning is that " whether the query contains ToMany joins (true/false)".

In case there is a legitimate worry that there might be someone in the userspace that does the (undocumented) due diligence and adds the DISTINCT keyword to their queries when they set `fetchJoinCollection` to false, and we don't want to break that code, then I can introduce a new `queryHasToManyJoins` instance variable, and use that instead of `fetchJoinCollection` (in which case, I would also need to find out what I need to default `fetchJoinCollection` to, but that would not be an unsurmountable problem).

Other that that, I admit that there is a potential future improvement, where `useOutputWalkers(forCountQuery: true)` could still returns false (i.e. saying that the "simple" count query should be used) even when `queryHasToManyJoins` is true, however `fetchJoinCollection` is false, and we know that a `COUNT(DISTINCT foo)` is going to be used.

So, are you ok for me to rename that instance variable, or would you rather that I should not touch it, and instead introduce a new instance variable?

Reply:
@d-ph I'm not saying that the literal meaning of `fetchJoinCollection` is correct. I'm saying that this is more complex because we actually care about a *different* thing depending on whether the DQL query is performing a distinct section or no. Here is my analysis

## counting queries

- all counting queries are incompatible with queries selecting multiple entities in the `FROM` (so the Paginator is actually incompatible with them).
- the fast counting query (tree-walker one) is incompatible with entities using a composite primary key
- the fast counting query is incompatible with queries using a `HAVING` clause
- the fast counting query is compatible with toMany joins when using `CountWalker::HINT_DISTINCT` as it will count distinct values (compatibility with `GROUP BY` clause is probably partial in the case of `toMany` join as I think it breaks if one entity can end up in multiple groups due to including a field of the `toMany` relation in the grouping). This is the reason why the Paginator sets this hint by default.
- the fast counting query is not compatible with `toMany` joins without `CountWalker::HINT_DISTINCT` (but this can go even faster for compatible queries)

## Result fetching

- the fast result fetching query (using a `limit` on the original query without changing it) is:
    - compatible with queries not using `toMany` joins
    - compatible with queries using `toMany` joins and `GROUP BY` when no fields from the grouping condition is dependant on the ToMany join (i.e. is part of the joined entity or of other entities joined again from that one)
    - incompatible with queries selecting data from `toMany` joins
    - compatible with queries doing `toMany` joins without selecting data from it (i.e. using it only for filtering in the WHERE clause or for aggregate functions) if the query does a `SELECT DISTINCT`
- walker-based fetching queries do not support entities using a composite primary key
- walker-based fetching queries do not support queries selecting multiple entities in the `FROM`
- the tree-walker fetching query does not support entities using a relation as primary key
- the tree walker fetching query does not support queries using a field from a `toMany` relation in its `ORDER BY` clause when using a limit (queries without a limit always use the fast result fetching query anyway)

In the existing code `$fetchJoinCollection` is all about deciding whether we can use the fast result fetching query vs using the more complex logic (which will then use either a tree walker or an output walker). This should be preferred when compatible as it is faster and is also the only result fetching mode compatible with composite primary keys.
I don't think renaming that field is helpful unless you can find a name that accurately corresponds to the unsupported cases (see the sublist). And I don't think we should replace it with multiple fields in the public API as that would be more complex for users than having a single boolean switching between the 2 modes. However, it might make sense to improve the documentation of supported cases (or even better to make the automatic guessing reliable enough so that they almost never need to configure it themselves anymore while getting the fastest compatible behavior).

Regarding output walker vs tree walker, the output walker supports more cases (assuming you don't already have an output walker as we can have only one, which is an uncommon case) but is generally less performant than then tree-walker query when both are supported.

Reply:
The internal guessing-related state might involve more properties containing the analysis of the query. But it could actually be simpler for it to only split the `useOutputWalkers` field into 2 separate fields (one for counting and one for result fetching) that can be configured by the guessing logic.

Reply:
That was a lot to take in for me.

After a substantial amount of thought, I would like to propose and get your opinion on the following.

1. Rename the `$fetchJoinCollection` to `$query(Potentially)ProducesDuplicateRows: bool`.
1.1. Why: The new name expresses the "root of the problem", and does not go into details about "how it comes about". It also succinctly expresses why pagination is not a trivial task.
1.2. What it achieves: it allows the end user to correctly express their confidence (or lack of it) in whether their query produces duplicate rows (which necessitate that the Paginator should run more complex queries). In particular: it puts the onus on the end user to ensure that this "declaration" (in particular: when they "declare false") is accurate.
1.3. What it achieves: In case the end user's query has a `ToMany` join, but it doesn't select any fields from it (or only selects SQL aggregate functions that use those fields), and the query uses a "SELECT DISTINCT" (or an appropriate GROUP BY), then the end user can succinctly inform the Paginator that their query is safe for the "fast result & count queries" without explaining the minute details about what exactly allows those fast queries to be used.
1.4. Bonus: It would "fix" what I said somewhere above that currently when `$fetchJoinCollection` is false, then the `toIterator()` runs the "fast result query", however there is nothing that indicates that the end user did the due diligence of using `SELECT DISTINCT`. The new name of `$queryProducesDuplicateRows=false` would in this case communicate the intended use case (and requirements) to the end user in a more complete way than `$fetchJoinCollection`.
1.5. Breaking change handling: the `getFetchJoinCollection()` getter would be preserved but deprecated. A new appropriate getter would be introduced and recommended as a replacement.
1.6. Why: This new name would resolve the review comment that was raised by you in relation to line 257 in the PR.
1.7. Side note: "$fetchJoinCollection" is just a terrible name, as it suggests that the Paginator would do something to the joined collections in the query.

2. Split `$useOutputWalkers` into `$useResultQueryOutputWalker` and `$useCountQueryOutputWalker`.
2.1. Why: Apart from what you said, splitting this config variable just makes sense. The two output walkers (i.e. on the result query and the count query) have different reasons to be enabled/disabled.
2.2. Breaking change handling: the `(get|set)UseOutputWalkers()` functions would be preserved but deprecated. A new set of functions (4 in total) would be introduced and recommended as a replacement.
2.3. Downside: There would effectively be one more pair of public functions to (optionally) configure the Paginator. But again: it should be viewed as a long overdue code change, rather than an questionable and optional addition.
2.4. Benefit: It would allow end users to manually control whether they are confident to run the "fast" count query, without affecting the runtime of the `toIterator()` function.

3. The auto-detection/guessing would investigate the given Query and appropriately set the variables mentioned above.
3.1. The auto-detection would initially not attempt to auto-detect whether a presence of a `ToMany` join is compatible with the "fast" queries. This is to limit the scope of this PR. It obviously could be attempted in future PRs. As such, as soon as there is a `ToMany` join detected, the "fast" queries would not be used (which is equivalent to the current (default) behaviour of the Paginator).

Q1. Are you personally ok with the proposition above?

---

> the fast counting query is compatible with toMany joins when using CountWalker::HINT_DISTINCT as it will count distinct values

Q2. Could you please confirm that you meant: `is compatible with toMany joins when using CountWalker::HINT_DISTINCT, AND when the query does NOT select any of the toMany entities/fields`? Because when the query selects any of the toMany fields, then it will make the CountWalker (i.e. the fast query) not work correctly, regardless of whether CountWalker::HINT_DISTINCT is added or not.

> when no fields from the grouping condition is dependant on the ToMany join (i.e. is part of the joined entity or of other entities joined again from that one)

Thanks for mentioning the case of `other entities joined again from that one`. This is quite important and I did not initially think of it.

> [the points about when the walker-based fetching queries and tree-walker fetching queries are supported]

That is good to know, however note that I do not intend to touch the region of the code that makes the decision whether the walker-based or tree-walker fetching query should be used.

> [The "fast" result query] is also the only result fetching mode compatible with composite primary keys.

Fair point, however this again is not something that I'm going to change or touch in my PR. Note that if a query uses composite primary keys AND produces duplicate rows (i.e. is not compatible with the "fast" result query), then the Paginator will not able to be used at all, and it's no different to the current situation in the Paginator.

> Regarding output walker vs tree walker, the output walker supports more cases (...) but is generally less performant than then tree-walker query when both are supported.

This again is good to know, however I don't intend to auto-detect/guess in this PR which of the two walkers should be used by the result-fetching query.

---

Could you reply to Q1 and Q2, please, and share any other thought you have about what I said above?

Reply:
Sorry, are you referring to the following diff in the current PR?
```php
    private function useOutputWalker(Query $query): bool
    {
        if ($this->useOutputWalkers === null) {
            return (bool) $query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) === false;
        }

        return $this->useOutputWalkers;
    }
```
->
```php
    private function useOutputWalker(Query $query, bool $forCountQuery = false): bool
    {
        if ($this->useOutputWalkers !== null) {
            return $this->useOutputWalkers;
        }

        // When a custom output walker already present, then do not use the Paginator's.
        if ($query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) !== false) {
            return false;
        }

        // When not joining onto *ToMany relations, then do not use the more complex CountOutputWalker.
        return ! (
            $forCountQuery
// (rest of the diff not relevant)
        );
    }
```
Do we both agree that this code diff does not result in any change to whether the walker-based or the tree-walker _result-fetching_ query should be used?

Or do you refer to the proposition of splitting `useOutputWalker` into two variables? In that case I can assure you that I do not intend to modify the current logic for the `$useResultQueryOutputWalker` variable. It could be done in future PRs, most likely by different contributors than me.

Note that I started this PR with only the intention of adding code that "if there are any dql-joins present, then use the fast version of the queries". I believe this alone would immediately address 80% of the use cases in the wild. Anything more than this, I consider a nicety. However, I do agree that going the extra mile and checking for *ToMany dql-joins should be easy enough, so as an exception from my original plan, I'm willing to attempt to make that change in the current PR. Anything other than that is a "less than 5% of use cases" matter in my mind, and although I do appreciate us discussing it, I do not intend to address it in this PR.

Reply:
when `$fetchJoinCollection` is `true` (the default), forbidding to use the fast result fetching query, `useOutputWalker` is what decides between fetching results with an output walker or a tree walker.

Reply:
> when `$fetchJoinCollection` is true (the default), forbidding to use the fast result fetching query,

I agree.

> when `$fetchJoinCollection` is true (the default) (...), useOutputWalker is what decides between fetching results with an output walker or a tree walker.

I agree.

> useOutputWalker is what decides between fetching results with an output walker or a tree walker.

Do we both agree that my code diff does not change the behaviour of `useOutputWalker`?

> when `$fetchJoinCollection` is true (the default)

Are you trying to say that because the new `Paginator::newWithAutoDetection()` function in this PR is now going to decide whether the (former) `$fetchJoinCollection` is true or false, it means that I should pay more attention to when the function decides that `$fetchJoinCollection` should be "set" to true, because this implies that `useOutputWalker` is going to be run? To this I would say that I partially see the concern here, however I don't view `Paginator::newWithAutoDetection()` function as deciding whether the `$fetchJoinCollection` should be set to true, but rather whether the `$fetchJoinCollection` should "revert to the current behaviour of the Paginator" (i.e. setting the variable to true). It's not ideal the the default behaviour of `Paginator::newWithAutoDetection()` could lead to problems in some cases, but likewise the current default behaviour of the Paginator class would lead to _the same problems_. And for this reason, I do not see why I should go even farther with this PR by making sure that the details about the output/tree walkers in the result-fetching query are taken into the consideration. Please correct me if I'm wrong.

If you're not referring to the consequences of the `Paginator::newWithAutoDetection()` function, then I truly struggle to see what we are discussing here, and I would appreciate some more help in understanding your standpoint. Why exactly do you believe I should concern myself with the the output/tree walkers in the result-fetching query in this PR?